src/kernels/composable_kernel/include/kernel_algorithm/gridwise_generic_reduction.hpp

```diff
@@ -0,0 +1,466 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef CK_GRIDWISE_GENERIC_REDUCTION_HPP
+#define CK_GRIDWISE_GENERIC_REDUCTION_HPP
+
+#include "float_type.hpp"
+#include "reduction_common.hpp"
+#include "reduction_operator.hpp"
+#include "reduction_kernel_simple_configurator.hpp"
+
+#include "tuple_ext.hpp"
+
+#include "gridwise_generic_2d_reduction_direct_threadwise.hpp"
+#include "gridwise_generic_2d_reduction_direct_warpwise.hpp"
+#include "gridwise_generic_2d_reduction_blockwise.hpp"
+#include "gridwise_generic_2d_reduction_multiblock.hpp"
+
+namespace ck {
+
+template <int BlkGroupSize,
+          int BlockSize,
+          typename srcDataType,  // the type with which the data of the source tensor are stored
+          typename dstDataType,  // the type with which the data of the destintion tensor are stored
+          typename compType,     // the type used by the reduce binary operator
+          typename srcDesc,      // the descriptor representing the source tensor to be reduced
+          typename toReduceDims, // the Sequence<...> consists of the indexes of toReduce dimensions
+                                 // in the source tensor descriptor
+          typename invariantDims, // the Sequence<...> consists of the indexes of invariant
+                                  // dimensions in the source tensor descriptor (can be empty)
+          typename dstDesc,  // the descriptor representing the destination tensor where the reduced
+                             // tensor data are saved/added
+          int op_I,          // the enumerate value representing the operation used in Reduction
+          int reduceImpl_I,  // the enumerate value representing the ReductionMethod
+          int nanPropaOpt_I, // the enumerate value representing the NanPropagation Option
+          int reduceIndicesOpt_I, // the enumerate value representing the Reduce Indices Option
+          int GredThreadBufferLength,
+          int GredAccessesPerThreadInBlock,
+          int GredAccessesPerThreadInWarp>
+struct Gridwise_generic_reduction
+{
+    static constexpr auto reduceImpl = static_cast<ckReductionMethod_t>(reduceImpl_I);
+    static constexpr bool is_method_multiblock =
+        (reduceImpl == CK_Reduce_MultiBlock) ? true : false;
+    static constexpr auto op          = static_cast<ckReduceTensorOp_t>(op_I);
+    static constexpr auto nanPropaOpt = static_cast<ckNanPropagation_t>(nanPropaOpt_I);
+    static constexpr auto reduceIndicesOpt =
+        static_cast<ckReduceTensorIndices_t>(reduceIndicesOpt_I);
+
+    template <ckReductionMethod_t impl, int callId>
+    struct Gridwise_generic_2d_reduction_wrapper;
+
+    // wrapper for switching to the Reduce_DirectThreadWise method
+    template <int callId>
+    struct Gridwise_generic_2d_reduction_wrapper<CK_Reduce_DirectThreadWise, callId>
```

Comment:
Please use the type name of enum ```ckReductionMethod_t``` before the CK_Reduce_DirectThreadWise. Without then enum type name, It is hard to tell if ```CK_Reduce_DirectThreadWise``` is an enum or not.

Please do this for all the enums

Reply:
Ok.  First-time to recognize this issue.   A in-advance statement like " template <ckReductionMethod_t impl, int callId>
    struct Gridwise_generic_2d_reduction_wrapper; "    is not enough ?

Reply:
sorry for confusion, I mean ```ckReductionMethod_t::CK_Reduce_DirectThreadWise```, this is more clear which type of enum ```CK_Reduce_DirectThreadWise``` is.

Otherwise, we would have a lot of enum and we wouldn't quickly understand what they are

Reply:
ping

Reply:
Check https://github.com/ROCmSoftwarePlatform/MIOpen/pull/322/commits/63eccb3a606d9d30b2c08fce8b7f42f72aecb40e